File: tools/lsp/parser/ast/ast.go
package ast

// Node is the base interface for all AST nodes.
type Node interface {
	NodeLine() int
	NodeColumn() int
}

// Definition is a top-level definition (workflow, activity, worker, or namespace).
type Definition interface {
	Node
	defNode()
}

// Statement is a statement inside a body.
type Statement interface {
	Node
	stmtNode()
}

// Pos holds source position information.
type Pos struct {
	Line   int
	Column int
}

func (p Pos) NodeLine() int   { return p.Line }
func (p Pos) NodeColumn() int { return p.Column }

// Ref is a named reference to another AST node, resolved after parsing.
type Ref[T any] struct {
	Pos
	Name     string
	Resolved T
}

// File represents a parsed .twf file.
type File struct {
	Definitions []Definition
}

// ---------------------------------------------------------------------------
// Top-level definitions
// ---------------------------------------------------------------------------

type WorkflowDef struct {
	Pos
	Name       string
	Params     string // opaque content inside parens
	ReturnType string // opaque, optional
	State      *StateBlock
	Signals    []*SignalDecl
	Queries    []*QueryDecl
	Updates    []*UpdateDecl
	Body       []Statement
}

func (*WorkflowDef) defNode() {}

type ActivityDef struct {
	Pos
	Name       string
	Params     string
	ReturnType string
	Body       []Statement
}

func (*ActivityDef) defNode() {}

type WorkerDef struct {
	Pos
	Name       string
	Workflows  []Ref[*WorkflowDef]
	Activities []Ref[*ActivityDef]
	Services   []Ref[*NexusServiceDef] // nexus service references
}

func (*WorkerDef) defNode() {}

// NamespaceWorker is a worker instantiation inside a namespace block.
type NamespaceWorker struct {
	Pos
	Worker  Ref[*WorkerDef]
	Options *OptionsBlock
}

// NamespaceEndpoint is a nexus endpoint instantiation inside a namespace block.
type NamespaceEndpoint struct {
	Pos
	EndpointName string
	Namespace    string // set by resolver: name of the owning namespace
	Options      *OptionsBlock
}

// NamespaceDef is a namespace definition that instantiates workers with options.
type NamespaceDef struct {
	Pos
	Name      string
	Workers   []NamespaceWorker
	Endpoints []NamespaceEndpoint
}

func (*NamespaceDef) defNode() {}

// ---------------------------------------------------------------------------
// Workflow-level declarations (embedded in WorkflowDef)
// ---------------------------------------------------------------------------

type SignalDecl struct {
	Pos
	Name   string
	Params string
	Body   []Statement // handler body
}

func (*SignalDecl) stmtNode() {}

type QueryDecl struct {
	Pos
	Name       string
	Params     string
	ReturnType string
	Body       []Statement // handler body (restricted: no temporal primitives)
}

func (*QueryDecl) stmtNode() {}

type UpdateDecl struct {
	Pos
	Name       string
	Params     string
	ReturnType string
	Body       []Statement // handler body
}

func (*UpdateDecl) stmtNode() {}

// ---------------------------------------------------------------------------
// Statements
// ---------------------------------------------------------------------------

type ActivityCall struct {
	Pos
	Activity Ref[*ActivityDef]
	Args     string
	Result   string // optional
	Options  *OptionsBlock
}

func (*ActivityCall) stmtNode() {}

// WorkflowCallMode describes how a workflow call is executed.
type WorkflowCallMode int

const (
	CallChild  WorkflowCallMode = iota // bare workflow call (child)
	CallDetach                         // detach workflow (fire-and-forget)
)

type WorkflowCall struct {
	Pos
	Mode     WorkflowCallMode
	Workflow Ref[*WorkflowDef]
	Args     string
	Result   string // optional
	Options  *OptionsBlock
}

func (*WorkflowCall) stmtNode() {}

// ---------------------------------------------------------------------------
// Async target discriminated union
// ---------------------------------------------------------------------------

// AsyncTarget is the interface for async operation targets used in await,
// await-one-case, and promise statements. Exactly one concrete type is used.
type AsyncTarget interface {
	asyncTarget() // unexported marker
}

// AsyncTargetKind returns a string identifying the kind of async target.
func AsyncTargetKind(t AsyncTarget) string {
	switch t.(type) {
	case *TimerTarget:
		return "timer"
	case *SignalTarget:
		return "signal"
	case *UpdateTarget:
		return "update"
	case *ActivityTarget:
		return "activity"
	case *WorkflowTarget:
		return "workflow"
	case *NexusTarget:
		return "nexus"
	case *IdentTarget:
		return "ident"
	default:
		return "unknown"
	}
}

type TimerTarget struct {
	Duration string
}

func (*TimerTarget) asyncTarget() {}

type SignalTarget struct {
	Signal Ref[*SignalDecl]
	Params string
}

func (*SignalTarget) asyncTarget() {}

type UpdateTarget struct {
	Update Ref[*UpdateDecl]
	Params string
}

func (*UpdateTarget) asyncTarget() {}

type ActivityTarget struct {
	Activity Ref[*ActivityDef]
	Args     string
	Result   string
}

func (*ActivityTarget) asyncTarget() {}

type WorkflowTarget struct {
	Workflow Ref[*WorkflowDef]
	Mode     WorkflowCallMode
	Args     string
	Result   string
}

func (*WorkflowTarget) asyncTarget() {}

type NexusTarget struct {
	Endpoint  Ref[*NamespaceEndpoint]
	Service   Ref[*NexusServiceDef]
	Operation Ref[*NexusOperation]
	Args      string
	Result    string
	Detach    bool
}

func (*NexusTarget) asyncTarget() {}

// IdentResolution holds the resolved target of an ident reference.
// Exactly one field is non-nil after successful resolution.
type IdentResolution struct {
	Promise   *PromiseStmt
	Condition *ConditionDecl
}

type IdentTarget struct {
	Name     string
	Result   string
	Resolved IdentResolution
}

func (*IdentTarget) asyncTarget() {}

// AwaitStmt represents a single await statement.
type AwaitStmt struct {
	Pos
	Target AsyncTarget
}

func (*AwaitStmt) stmtNode() {}

// AwaitAllBlock represents an "await all:" block that waits for all operations to complete.
type AwaitAllBlock struct {
	Pos
	Body []Statement
}

func (*AwaitAllBlock) stmtNode() {}

// AwaitOneCase represents a single case in an "await one:" block.
// Can be signal, update, timer, activity, workflow, nexus, ident, or nested await all.
type AwaitOneCase struct {
	Pos
	Target   AsyncTarget    // nil when AwaitAll is set
	AwaitAll *AwaitAllBlock // nil when Target is set
	Body     []Statement
}

func (*AwaitOneCase) stmtNode() {}

// AwaitOneBlock represents an "await one:" block that waits for the first case to complete.
type AwaitOneBlock struct {
	Pos
	Cases []*AwaitOneCase
}

func (*AwaitOneBlock) stmtNode() {}

// SwitchCase represents a single case in a switch block.
type SwitchCase struct {
	Pos
	Value string // opaque expression after "case"
	Body  []Statement
}

type SwitchBlock struct {
	Pos
	Expr    string // opaque, paren-delimited
	Cases   []*SwitchCase
	Default []Statement // optional else block
}

func (*SwitchBlock) stmtNode() {}

type IfStmt struct {
	Pos
	Condition string // opaque, paren-delimited
	Body      []Statement
	ElseBody  []Statement // optional
}

func (*IfStmt) stmtNode() {}

// ForVariant describes the kind of for loop.
type ForVariant int

const (
	ForInfinite    ForVariant = iota // for:
	ForConditional                   // for (condition):
	ForIteration                     // for (var in collection):
)

type ForStmt struct {
	Pos
	Variant   ForVariant
	Condition string // for conditional loops
	Variable  string // for iteration loops
	Iterable  string // for iteration loops
	Body      []Statement
}

func (*ForStmt) stmtNode() {}

type ReturnStmt struct {
	Pos
	Value string // opaque, optional
}

func (*ReturnStmt) stmtNode() {}

// CloseReason classifies the kind of workflow close.
type CloseReason int

const (
	CloseComplete      CloseReason = iota // close complete
	CloseFailWorkflow                     // close fail
	CloseContinueAsNew                    // close continue_as_new
)

type CloseStmt struct {
	Pos
	Reason CloseReason
	Args   string // opaque, optional (parenthesized args)
}

func (*CloseStmt) stmtNode() {}

type BreakStmt struct {
	Pos
}

func (*BreakStmt) stmtNode() {}

type ContinueStmt struct {
	Pos
}

func (*ContinueStmt) stmtNode() {}

type RawStmt struct {
	Pos
	Text string
}

func (*RawStmt) stmtNode() {}

type Comment struct {
	Pos
	Text string
}

func (*Comment) stmtNode() {}

// ---------------------------------------------------------------------------
// State block and new primitives
// ---------------------------------------------------------------------------

// StateBlock represents a state: block at the top of a workflow definition.
type StateBlock struct {
	Pos
	Conditions []*ConditionDecl
	RawStmts   []*RawStmt
}

// ConditionDecl represents a condition declaration inside a state block.
type ConditionDecl struct {
	Pos
	Name string
}

// PromiseStmt represents a promise declaration: promise name <- async_target
type PromiseStmt struct {
	Pos
	Name   string
	Target AsyncTarget
}

func (*PromiseStmt) stmtNode() {}

// SetStmt represents: set conditionName
type SetStmt struct {
	Pos
	Condition Ref[*ConditionDecl]
}

func (*SetStmt) stmtNode() {}

// UnsetStmt represents: unset conditionName
type UnsetStmt struct {
	Pos
	Condition Ref[*ConditionDecl]
}

func (*UnsetStmt) stmtNode() {}

// ---------------------------------------------------------------------------
// Nexus definitions and calls
// ---------------------------------------------------------------------------

// NexusOperationType distinguishes async vs sync nexus operations.
type NexusOperationType int

const (
	NexusOpAsync NexusOperationType = iota
	NexusOpSync
)

// NexusOperation is an operation inside a nexus service definition.
type NexusOperation struct {
	Pos
	OpType       NexusOperationType
	Name         string
	Workflow      Ref[*WorkflowDef] // async only: backing workflow
	Params       string      // sync only
	ReturnType   string      // sync only
	Body         []Statement // sync only
}

// NexusServiceDef is a top-level nexus service definition.
type NexusServiceDef struct {
	Pos
	Name       string
	Operations []*NexusOperation
}

func (*NexusServiceDef) defNode() {}

// NexusCall is a nexus service operation call inside a workflow body.
type NexusCall struct {
	Pos
	Detach    bool
	Endpoint  Ref[*NamespaceEndpoint]
	Service   Ref[*NexusServiceDef]
	Operation Ref[*NexusOperation]
	Args      string
	Result    string // optional
	Options   *OptionsBlock
}

func (*NexusCall) stmtNode() {}

// ---------------------------------------------------------------------------
// Options blocks
// ---------------------------------------------------------------------------

// OptionsBlock represents a structured options { ... } block.
type OptionsBlock struct {
	Pos
	Entries []*OptionEntry
}

// OptionEntry represents a single key-value pair or nested block inside options.
type OptionEntry struct {
	Pos
	Key       string
	Value     string         // literal for flat entries
	ValueType string         // "string", "duration", "number", "bool", "enum"
	Nested    []*OptionEntry // non-nil for nested blocks (e.g. retry_policy)
}

File: tools/visualizer/src/types/ast.ts
// TypeScript types mirroring the Go AST JSON output

export interface Position {
  line: number
  column: number
}

// Per-file parse error
export interface FileError {
  file: string
  error: string
  stderr?: string
}

// Top-level file
export interface TWFFile {
  definitions: Definition[]
  // Added for focused-file visualization
  focusedFile?: string
  // Per-file parse errors and warnings
  errors?: FileError[]
}

// Definition types
export type Definition = WorkflowDef | ActivityDef

export interface WorkflowDef extends Position {
  type: 'workflowDef'
  name: string
  params: string
  returnType?: string
  options?: string
  state?: StateBlock
  signals: SignalDecl[]
  queries: QueryDecl[]
  updates: UpdateDecl[]
  body: Statement[]
  // Source file path (added by extension)
  sourceFile?: string
}

// State block declared at the top of a workflow definition
export interface StateBlock {
  conditions?: ConditionDecl[]
  rawStmts?: RawStmt[]
}

export interface ConditionDecl extends Position {
  name: string
}

export interface ActivityDef extends Position {
  type: 'activityDef'
  name: string
  params: string
  returnType?: string
  options?: string
  body: Statement[]
  // Source file path (added by extension)
  sourceFile?: string
}

// Handler declaration union (signal, query, update)
export type HandlerDecl = SignalDecl | QueryDecl | UpdateDecl

// Declaration types (with handler bodies)
export interface SignalDecl extends Position {
  type: 'signalDecl'
  name: string
  params: string
  body?: Statement[]
}

export interface QueryDecl extends Position {
  type: 'queryDecl'
  name: string
  params: string
  returnType?: string
  body?: Statement[]
}

export interface UpdateDecl extends Position {
  type: 'updateDecl'
  name: string
  params: string
  returnType?: string
  body?: Statement[]
}

// Statement types
export type Statement =
  | ActivityCall
  | WorkflowCall
  | AwaitStmt
  | AwaitAllBlock
  | AwaitOneBlock
  | SwitchBlock
  | IfStmt
  | ForStmt
  | ReturnStmt
  | CloseStmt
  | BreakStmt
  | ContinueStmt
  | RawStmt
  | Comment
  | PromiseStmt
  | SetStmt
  | UnsetStmt

export interface ActivityCall extends Position {
  type: 'activityCall'
  name: string
  args: string
  result?: string
  options?: string
}

export type WorkflowCallMode = 'child' | 'detach'

export interface WorkflowCall extends Position {
  type: 'workflowCall'
  mode: WorkflowCallMode
  namespace?: string
  name: string
  args: string
  result?: string
  options?: string
}

// Single await statement: await timer/signal/update/activity/workflow/ident
export type AwaitStmtKind = 'timer' | 'signal' | 'update' | 'activity' | 'workflow' | 'ident'

export interface AwaitStmt extends Position {
  type: 'await'
  kind: AwaitStmtKind
  timer?: string
  signal?: string
  signalParams?: string
  update?: string
  updateParams?: string
  activity?: string
  activityArgs?: string
  activityResult?: string
  workflow?: string
  workflowMode?: string
  workflowNamespace?: string
  workflowArgs?: string
  workflowResult?: string
  // Ident await (promise or condition reference)
  ident?: string
  identResult?: string
}

// await all: waits for all operations to complete
export interface AwaitAllBlock extends Position {
  type: 'awaitAll'
  body: Statement[]
}

// await one case: signal, update, timer, activity, workflow, nested await all, or ident
export type AwaitOneCaseKind = 'signal' | 'update' | 'timer' | 'activity' | 'workflow' | 'await_all' | 'ident'

export interface AwaitOneCase extends Position {
  kind: AwaitOneCaseKind
  // Signal case
  signal?: string
  signalParams?: string
  // Update case
  update?: string
  updateParams?: string
  // Timer case
  timer?: string
  // Activity case
  activity?: string
  activityArgs?: string
  activityResult?: string
  // Workflow case
  workflow?: string
  workflowMode?: string
  workflowNamespace?: string
  workflowArgs?: string
  workflowResult?: string
  // Await all case (nested)
  awaitAll?: AwaitAllBlock
  // Ident case (promise or condition reference)
  ident?: string
  identResult?: string
  // Body executed when this case wins (optional - can be empty)
  body: Statement[]
}

// await one: waits for first case to complete
export interface AwaitOneBlock extends Position {
  type: 'awaitOne'
  cases: AwaitOneCase[]
}

export interface SwitchCase extends Position {
  value: string
  body: Statement[]
}

export interface SwitchBlock extends Position {
  type: 'switch'
  expr: string
  cases: SwitchCase[]
  default?: Statement[]
}

export interface IfStmt extends Position {
  type: 'if'
  condition: string
  body: Statement[]
  elseBody?: Statement[]
}

export type ForVariant = 'infinite' | 'conditional' | 'iteration'

export interface ForStmt extends Position {
  type: 'for'
  variant: ForVariant
  condition?: string
  variable?: string
  iterable?: string
  body: Statement[]
}

export interface ReturnStmt extends Position {
  type: 'return'
  value?: string
}

export interface CloseStmt extends Position {
  type: 'close'
  reason: string // 'complete', 'fail', or 'continue_as_new'
  args?: string
}

export interface BreakStmt extends Position {
  type: 'break'
}

export interface ContinueStmt extends Position {
  type: 'continue'
}

export interface RawStmt extends Position {
  type: 'raw'
  text: string
}

export interface Comment extends Position {
  type: 'comment'
  text: string
}

// Promise statement: promise name <- async_target
export interface PromiseStmt extends Position {
  type: 'promise'
  name: string
  // Async target (exactly one set)
  timer?: string
  signal?: string
  signalParams?: string
  update?: string
  updateParams?: string
  activity?: string
  activityArgs?: string
  workflow?: string
  workflowNamespace?: string
  workflowArgs?: string
}

// Set a condition to true
export interface SetStmt extends Position {
  type: 'set'
  name: string
}

// Set a condition to false
export interface UnsetStmt extends Position {
  type: 'unset'
  name: string
}

// Type guards
export function isWorkflowDef(def: Definition): def is WorkflowDef {
  return def.type === 'workflowDef'
}

export function isActivityDef(def: Definition): def is ActivityDef {
  return def.type === 'activityDef'
}
